pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,220 @@
+		gitClient:              gitClient,
+		applicationLister:      appLister,
+		config:                 cfg,
+		gracePeriod:            gracePeriod,
+		lastFetchedCommitCache: cache,
+		logger:                 logger.Named("app-config-reporter"),
+	}, nil
+}
+
+func (r *Reporter) Run(ctx context.Context) error {
+	r.logger.Info("start running app-config-reporter")
+
+	// Pre-clone to cache the registered git repositories.
+	r.gitRepos = make(map[string]git.Repo, len(r.config.Repositories))
+	for _, repoCfg := range r.config.Repositories {
+		repo, err := r.gitClient.Clone(ctx, repoCfg.RepoID, repoCfg.Remote, repoCfg.Branch, "")
+		if err != nil {
+			r.logger.Error("failed to clone repository",
+				zap.String("repo-id", repoCfg.RepoID),
+				zap.Error(err),
+			)
+			return err
+		}
+		r.gitRepos[repoCfg.RepoID] = repo
+	}
+
+	// FIXME: Think about sync interval of app config reporter
+	ticker := time.NewTicker(r.config.SyncInterval.Duration())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := r.checkApps(ctx); err != nil {
+				r.logger.Error("failed to check application configurations defined in Git", zap.Error(err))
+			}
+		case <-ctx.Done():
+			r.logger.Info("app-config-reporter has been stopped")
+			return nil
+		}
+	}
+}
+
+// checkApps checks and reports two types of applications.
+// One is applications registered in Control-plane already, and another is ones that aren't registered yet.
+func (r *Reporter) checkApps(ctx context.Context) (err error) {
+	if len(r.gitRepos) == 0 {
+		r.logger.Info("no repositories were configured for this piped")
+		return
+	}
+
+	var (
+		unusedApps      = make([]*pipedservice.ApplicationConfiguration, 0)
+		appsToBeUpdated = make([]*pipedservice.ApplicationConfiguration, 0)
+		appsMap         = r.listApplications()
+	)
+	for repoID, repo := range r.gitRepos {
+		if err = repo.Pull(ctx, repo.GetClonedBranch()); err != nil {
+			r.logger.Error("failed to update repo to latest",
+				zap.String("repo-id", repoID),
+				zap.Error(err),
+			)
+			return
+		}
+
+		// TODO: Collect unused application configurations that aren't used yet
+		//   Currently, it could be thought the best to open files that suffixed by .pipe.yaml
+
+		var headCommit git.Commit
+		// Get the head commit of the repository.
+		headCommit, err = repo.GetLatestCommit(ctx)
+		if err != nil {
+			return
+		}
+		var lastFetchedCommit interface{}
+		lastFetchedCommit, err = r.lastFetchedCommitCache.Get(repoID)
+		if err != nil && !errors.Is(err, cache.ErrNotFound) {
+			r.logger.Error("failed to get the last fetched commit from cache", zap.Error(err))
+		}
+		if headCommit.Hash == lastFetchedCommit.(string) {
+			continue
+		}
+		apps, ok := appsMap[repoID]
+		if !ok {
+			continue
+		}
+		for _, app := range apps {
+			gitPath := app.GetGitPath()
+			_ = filepath.Join(repo.GetPath(), gitPath.Path, gitPath.ConfigFilename)
+			// TODO: Collect applications that need to be updated
+		}
```

Comment:
Instead of iterate apps to check like this, we can use `git diff previousHash currentHash --name-only` to have a list of changed files. And then find the updated ones and un-registered ones based on that list.